crates/flux-fhir-analysis/src/wf/sortck.rs

```diff
@@ -270,8 +263,58 @@ impl<'genv, 'tcx> InferCtxt<'genv, 'tcx> {
                     _ => Err(self.emit_field_not_found(&sort, *fld)),
                 }
             }
-            // (VR): Pretty sure we don't synthesize these?
-            _ => span_bug!(expr.span, "unexpected expr `{expr:?}` in synth expr"),
+            fhir::ExprKind::Constructor(Some(path), field_exprs, spread) => {
+                // if we have S then we can synthesize otherwise we fail
+                // first get the sort based on the path - for example S { ... } => S
+                // and we should expect sort to be a struct or enum app
+                let path_def_id = match path.res {
+                    ExprRes::Struct(def_id) | ExprRes::Enum(def_id) => def_id,
+                    _ => span_bug!(expr.span, "unexpected path in constructor"),
+                };
+                let sort_def = match self.genv.adt_sort_def_of(path_def_id) {
+                    Ok(sd) => sd,
+                    Err(_) => span_bug!(path.span, "Unexpected path in constructor"),
+                };
+                // generate fresh inferred sorts for each param
+                let fresh_args = (0..sort_def.param_count())
+                    .map(|_| self.next_sort_var())
+                    .collect_vec();
+                let sort_by_name = sort_def.sort_by_field_name(&fresh_args);
+                let sort = rty::Sort::App(rty::SortCtor::Adt(sort_def.clone()), fresh_args.into());
+                let mut used_fields = FxHashSet::default();
+                for expr in field_exprs.iter() {
+                    // check each expression against the sort
+                    // which will unify inferred variables
+                    if let Some(sort) = sort_by_name.get(&expr.ident.name) {
+                        self.check_expr(&expr.expr, sort)?;
+                    } else {
+                        return Err(self.emit_field_not_found(&sort, expr.ident));
+                    }
+                    if let Some(old_field) = used_fields.replace(expr.ident) {
```

Comment:
@nilehmann - I'm kinda annoyed that we have to do this in both checking and synthesis. It feels like this shouldn't really be here but I'm not sure there's a way around it.

Reply:
yes, I agree. We could move part of the logic to another function that only checks if all fields are present.

```rust
fn validate_constructor_fields(&self, sort_def: &AdtSortDef, exprs: &[fhir::FieldExpr], has_spread: bool);
```